server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java

```diff
@@ -1108,4 +1038,160 @@ public static ClusterBlock createUUIDBasedBlock(ClusterBlock clusterBlock) {
+                    "opening indices [{}]",
+                    String.join(",", indicesToOpen.stream().map(i -> (CharSequence) i.getIndex().toString())::iterator)
+                )
+            );
+
+            final Metadata.Builder metadata = Metadata.builder(currentState.metadata());
+            final ClusterBlocks.Builder blocks = ClusterBlocks.builder().blocks(currentState.blocks());
+            final Version minIndexCompatibilityVersion = currentState.getNodes().getMaxNodeVersion().minimumIndexCompatibilityVersion();
+
+            for (IndexMetadata indexMetadata : indicesToOpen) {
+                final Index index = indexMetadata.getIndex();
+                if (indexMetadata.getState() != IndexMetadata.State.OPEN) {
+                    final Settings.Builder updatedSettings = Settings.builder().put(indexMetadata.getSettings());
+                    updatedSettings.remove(VERIFIED_BEFORE_CLOSE_SETTING.getKey());
+
+                    IndexMetadata newIndexMetadata = IndexMetadata.builder(indexMetadata)
+                        .state(IndexMetadata.State.OPEN)
+                        .settingsVersion(indexMetadata.getSettingsVersion() + 1)
+                        .settings(updatedSettings)
+                        .timestampRange(IndexLongFieldRange.NO_SHARDS)
+                        .build();
+
+                    // The index might be closed because we couldn't import it due to an old incompatible
+                    // version, so we need to verify its compatibility.
+                    newIndexMetadata = indexMetadataVerifier.verifyIndexMetadata(newIndexMetadata, minIndexCompatibilityVersion);
+                    try {
+                        indicesService.verifyIndexMetadata(newIndexMetadata, newIndexMetadata);
+                    } catch (Exception e) {
+                        throw new ElasticsearchException("Failed to verify index " + index, e);
+                    }
+                    metadata.put(newIndexMetadata, true);
+                }
+
+                // Always removes index closed blocks (note: this can fail on-going close index actions)
+                blocks.removeIndexBlockWithId(index.getName(), INDEX_CLOSED_BLOCK_ID);
+            }
+
+            ClusterState updatedState = ClusterState.builder(currentState).metadata(metadata).blocks(blocks).build();
+
+            final RoutingTable.Builder routingTable = RoutingTable.builder(updatedState.routingTable());
+            for (IndexMetadata previousIndexMetadata : indicesToOpen) {
+                if (previousIndexMetadata.getState() != IndexMetadata.State.OPEN) {
+                    routingTable.addAsFromCloseToOpen(updatedState.metadata().getIndexSafe(previousIndexMetadata.getIndex()));
+                }
+            }
+            return ClusterState.builder(updatedState).routingTable(routingTable.build()).build();
+        }
+    }
+
+    public static class OpenIndicesTask implements ClusterStateTaskListener, ClusterStateAckListener {
```

Comment:
Also here

```suggestion
    private static class OpenIndicesTask implements ClusterStateTaskListener, ClusterStateAckListener {
```

May as well make it a `record` too.

Reply:
👍, b23ecddb9c63ea1504ffc897350e86154eaffe53 and f584fab22c0